primitives/arithmetic/src/lib.rs

```diff
@@ -85,8 +86,328 @@ where
+///
+/// This implementation is currently in a balanced position between being performant and accurate.
+///
+/// 1. We prefer storing original indices, and sorting the `input` only once. This will save the
+///    cost of sorting per round at the cost of a little bit of memory.
+/// 2. The granularity of increment/decrements is determined by the difference and the number of
+///    elements in `input` and their sum difference with `targeted_sum`. In short, the
+///    implementation is very likely to over-increment/decrement in case the `diff / input.len()` is
+///    rather small.
+///
+/// This function can return an error is if `T` cannot be built from the size of `input`, or if
+/// `sum(input)` cannot fit inside `T`. Moreover, if any of the internal operations saturate, it
+/// will also return en `Err`.
+///
+/// Based on the above, the best use case of the function will be only to correct small rounding
+/// errors. If the difference of the elements is vert large, then the subtraction/addition of one
+/// element with `diff / input.len()` might easily saturate and results will be `Err`.
+pub fn normalize<T>(input: &[T], targeted_sum: T) -> Result<Vec<T>, &'static str>
+	where T: Clone + Copy + Ord + BaseArithmetic + Debug,
+{
+	let mut sum = T::zero();
+	for t in input.iter() {
+		sum = sum.checked_add(t).ok_or("sum of input cannot fit in `T`")?;
+	}
+	let count = input.len();
+
+	// early exit if needed.
+	if count.is_zero() {
+		return Ok(Vec::<T>::new());
+	}
+
+	let diff = targeted_sum.max(sum) - targeted_sum.min(sum);
+	if diff.is_zero() {
+		return Ok(input.to_vec());
+	}
+
+	let needs_bump = targeted_sum > sum;
+	let per_round = diff / count.try_into().map_err(|_| "failed to build T from usize")?;
+	let mut leftover = diff % count.try_into().map_err(|_| "failed to build T from usize")?;
+
+	let mut output_with_idx = input.iter().cloned().enumerate().collect::<Vec<(usize, T)>>();
+	// sort output once based on diff. This will require more data transfer and saving original
+	// index, but we sort only twice instead: once now and once at the very end.
+	output_with_idx.sort_unstable_by_key(|x| x.1);
+
+	if needs_bump {
+		// must increase the values a bit. Bump from the min element. Index of minimum is now zero
+		// because we did a sort. If at any point the min goes greater or equal the `max_threshold`,
+		// we move to the next minimum.
+		let mut min_index = 0;
+		// at this threshold we move to next index.
+		let threshold = output_with_idx
+			.last()
+			.expect("length of input is greater than zero; it must have a last; qed").1;
+
+		if !per_round.is_zero() {
+			for _ in 0..count {
+				output_with_idx[min_index].1 = output_with_idx[min_index].1
+					.checked_add(&per_round)
+					.ok_or("Failed to add.")?;
+				if output_with_idx[min_index].1 >= threshold {
+					min_index += 1;
+					min_index = min_index % count;
+				}
+			}
+		}
+
+		// continue with the previous min_index
+		while !leftover.is_zero() {
```

Comment:
```
let start = ...; // start index
let n_one_extra = leftover % count;
let base_amount = leftover / count;
for i in 0..n_one_extra {
    let idx = start + i % count;
    output[idx] += base_amount + 1;
}
for i in n_one_extra..count {
    let idx = start + i % count;
    output[idx] += base_amount + 1;
}
```

Reply:
Note that this would apply an equivalent (at most off by one) offset to all items rather than pushing some items.

Reply:
I will try to incorporate this for the `need_bump == true`, the else branch cannot use this as we want to dynamically tweak `leftover`.

Reply:
actually, if you agree, I'd like to keep the implementation as is, so that both branches have a similar logic. The main problem that we have and could cause the second loop to be very slow is what we have in the `else` part, which cannot be addressed by this.

Reply:
Not sure if I understand, but I think you could remove the `mut` of leftover with the above approach _and_ apply it to both branches, at the cost of change in logic (as described in https://github.com/paritytech/substrate/pull/6374/files/cb9c22f34d43da5f4f10431ba564d6480f648d49#r443487807 ) - but if that is not intended leave as is :) feel free to resolve this

Reply:
resolved offline.